core/src/test/scala/integration/kafka/api/BaseProducerSendTest.scala

```diff
@@ -276,20 +272,26 @@ abstract class BaseProducerSendTest extends KafkaServerTestHarness {
         assertEquals(partition, recordMetadata.partition)
       }
 
-      val leader1 = leaders(partition)
+      consumer1.subscribe(List(topic))
+
       // make sure the fetched messages also respect the partitioning and ordering
-      val fetchResponse1 = if (leader1.get == configs.head.brokerId) {
-        consumer1.fetch(new FetchRequestBuilder().addFetch(topic, partition, 0, Int.MaxValue).build())
-      } else {
-        consumer2.fetch(new FetchRequestBuilder().addFetch(topic, partition, 0, Int.MaxValue).build())
+      val records = new ArrayList[ConsumerRecord[Array[Byte], Array[Byte]]]()
+      val maxIters = numRecords * 50
+      var iters = 0
+      while (records.size < numRecords) {
+        for (record <- consumer1.poll(50)) {
+          records.add(record)
+        }
+        if (iters > maxIters)
+          throw new IllegalStateException("Failed to consume the expected records after " + iters + " iterations.")
+        iters += 1
```

Comment:
A more concise and better way to do this is:

```scala
TestUtils.waitUntilTrue(() => {
      for (record <- consumer.poll(50).asScala)
        records.add(record)
      records.size == numRecords
    }, "Failed to receive all expected records from the consumer")
```

Probably worth creating a helper method and reusing it.